File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""
In this module we create a class to define a Rectangle
"""


class Rectangle:
    """
    In this class we dfine a Rectangle
    """
    def __init__(self, width=0, height=0):
        self.width = width
        self.height = height

    @property
    def width(self):
        """
        gets the value of the width
        """
        return self.__width

    @width.setter
    def width(self, value):
        """
        sets the value of the width
        """
        if isinstance(value, int):
            if value >= 0:
                self.__width = value
            else:
                raise ValueError("width must be >= 0")
        else:
            raise TypeError("width must be an integer")

    @property
    def height(self):
        """
        gets the value of the height
        """
        return self.__height

    @height.setter
    def height(self, value):
        """
        sets the value of the height
        """
        if isinstance(value, int):
            if value >= 0:
                self.__height = value
            else:
                raise ValueError("height must be >= 0")
        else:
            raise TypeError("height must be an integer")

    def area(self):
        """
        returns the area of a Rectangle object
        """
        return (self.__width * self.__height)

    def perimeter(self):
        """
        returns the perimeter of a Rectangle object
        """
        if self.__width == 0 or self.__height == 0:
            return 0
        return 2 * (self.__width + self.__height)

    def __str__(self):
        """
        returns a simple representation of an object the frome
        """
        sheap = ''
        if self.__width == 0 or self.height == 0:
            return sheap
        for i in range(self.__height):
            if i != self.__height - 1:
                sheap += "#" * self.__width + "\n"
            else:
                sheap += "#" * self.__width
        return sheap

    def __repr__(self):
        """
        returns a simple representation of an object
        """
        return "Rectangle({}, {})".format(self.__width, self.__height)


if __name__ == "__main__":
    my_rectangle = Rectangle(2, 4)
    print(str(my_rectangle))
    print("--")
    print(my_rectangle)
    print("--")
    print(repr(my_rectangle))
    print("--")
    print(hex(id(my_rectangle)))
    print("--")
    # create new instance based on representation

    new_rectangle = eval(repr(my_rectangle))
    print(str(new_rectangle))
    print("--")
    print(new_rectangle)
    print("--")
    print(repr(new_rectangle))
    print("--")
    print(hex(id(new_rectangle)))
    print("--")

    print(new_rectangle is my_rectangle)
    print(type(new_rectangle) is type(my_rectangle))

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
// This class describe a rectangle

class Rectangle {
	constructor(w, h) {
		if (isNaN(w) || isNaN(h) || w <= 0 || h <= 0)
		{
		} else {
			this.width = w;
			this.height = h;
		}
	}
	print() {
		let row = '';
		let i;
		for (i = 0; i < this.width; i++)
			row += 'X';
		for (i = 0; i < this.height; i++)
			console.log(row);
	}
	rotate() {
		let row = '';
		let i;
		for (i = 0; i < this.height; i++)
			row += 'X';
		for (i = 0; i< this.width; i++)
			console.log(row);
	}
	double() {
		this.width = this.width * 2;
		this.height = this.height * 2;
	}
}
module.exports = Rectangle;
